core/ir/riscv64/codec.py

```diff
@@ -0,0 +1,1063 @@
+                flds = tokens[2].strip().split(' ')
+                mask = 0
+                match = 0
+                shift = len(tokens[3])
+                if name.startswith("c.") or name == "unimp":
+                    if shift != 16:
+                        err(
+                            f"Invalid compressed instruction size: {name} {shift} != 16")
+                        return False
+                elif shift != 32:
+                    err(
+                        f"Invalid uncompressed instruction size: {name} {shift} != 32")
+                    return False
+                shift -= 1
+                for c in tokens[3]:
+                    if c == '.':
+                        mask |= 0 << shift
+                        match |= 0 << shift
+                    elif c == '1':
+                        mask |= 1 << shift
+                        match |= 1 << shift
+                    elif c == '0':
+                        mask |= 1 << shift
+                        match |= 0 << shift
+                    else:
+                        err(f"invalid mask for {name}: {tokens[3]}")
+                        return False
+                    shift -= 1
+                self.instructions.append(
+                    Instruction(name, fmt, mask, match, flds, ext))
+
+        return True
+
+    def __sanity_check(self) -> bool:
+        '''
+        Ensure there are no duplicates among uncompressed instructions.
+
+        Note that compressed instructions can be duplicated and require
+        validation taking the ISA + operand levels in mind.
+        '''
+        inst_dict = {}
+        for i in self.instructions:
+            id = inst_dict.get(i.match)
+            if id and not i.is_compressed() and id.mask == i.mask:
+                err(f'Duplicate instruction: {i}')
+                return False
+            inst_dict[i.match] = i
+        return True
+
+    def parse_isl(self, isl_path) -> bool:
+        '''
+        Parse Instruction Set Listing files into the generator object.
+
+        Iterates over all .txt files in a given path assuming each file contains
+        instructions from a single ISA extension (or base ISA).
+
+        Returns true if parsing of all files succeeded.
+        '''
+        p = Path(isl_path)
+        res = True
+        for f in [f for f in p.iterdir() if f.name.endswith(".txt")]:
+            res = res and self.__parse_isl_file(f)
+            if not res:
+                break
+        self.__fixup_instructions()
+        return res and self.__sanity_check()
+
+    def generate_opcodes(self, template_file, out_file) -> bool:
+        '''
+        Generate the opcode_api.h header file.
+
+        Returns true if the output file was written successfully.
+        '''
+        tmpl_fld = '@OPCODES@'
+        if len(self.instructions) == 0:
+            return False
+        found_template_fld = False
+        idx = -1
+        # Replace tmpl_fld in the template_file and save as out_file
```

Comment:
nit: Our style guide says to prefer including end punctuation with complete sentences in comments -- here and elsewhere

Reply:
Done. I think I've caught all.